### ISSUE TITLE
'Element Visibility' trigger requires 'Observe DOM changes' to detect presence of 'hidden' attribute

### ISSUE THREAD

**Issue by kslstn** [matomo-org/tag-manager#458]
I noticed that the removal of the `hidden` attribute from an element wasn't picked up by my 'Element Visibility' trigger. Trying to debug by setup, I think I found that the issue seems to be at Tag Manager.

**Expected behavior**: given an element matching the selection method and having a `hidden` HTML attribute, the 'Element Visibility' trigger should by applied when the `hidden` attribute is removed or set to `false`.

**Observed behavior**: MTM ignores the `hidden` attribute.

**How to reproduce this**:
- Create a trigger for 'Element Visibility' and as selection method `Element ID` =  test.
- Run preview
- Using the inspector, add a DOM node: <div 'id="test" hidden></div> -> The trigger is actived
- Set the `hidden` attribute to `false` -> The trigger is not activated
- Remove the `hidden` attribute -> The trigger is not activated

**Comment by AltamashShaikh** [comment]
Hello @kslstn 

I tried reproducing the same

- Added new DOM Element `<div id="test-hidden" hidden>1234555</div>` on the test website.
- Created 1 New Tag and trigger(Element Visibility Trigger).

Noticed following things
- Set the `hidden` attribute to `false` --> This will not work since its a boolean and `hidden` does not have a value, to make it work you have you can try removing the `hidden` attribute using this code `document.getElementById("test-hidden").removeAttribute("hidden")`
- It is firing the tag when I remove the attribute `hidden` and scroll to the element.

**Comment by kslstn** [comment]
Ah thanks, that works!

A few comments on this though:
1. When a node is added with the `hidden` attribute, the trigger is applied (and again when the attribute is removed).
2. The UI says "Note: This might slow down the website if there are lot of DOM changes". As 'a lot' is unspecific (2 per second? 3000 per second?) I was perhaps overly cautious not enabling it.
3. If removing the `hidden` attribute is considered a DOM change, I'd consider removing a `style` attribute a DOM change too. But the trigger is applied when `style="display: none;"` is removed, even without 'Observe DOM changes'.

**Comment by AltamashShaikh** [comment]
@kslstn `Observe DOM changes` is required only when you are adding the element dynamically.

1. The ElementVisibility trigger is executed when the specified element is made visible
2.  Its for sites which has too many dynamic changes like on scroll new Ads are added to the DOM with new element, you don' need to worry if there are very few DOM changes happening in your website
3. Can you share an example for point no 3 ? The visibility trigger will fire when the specified element is visible 
Doing something like this will trigger `document.getElementById("test-hidden").removeAttribute("style")` visibility trigger.

**Comment by kslstn** [comment]
Thanks good to know about 2)!

Regarding 3), what I consider inconsistent behavior is that while 'Observe DOM changes' is not enabled:
1. Remove the attribute `style="display: none;"`: the trigger picks it up
2. Remove the attribute `hidden`: the trigger ignores it

I tested this simply by editing the DOM of one of my pages in the console.

With the help in the answers above, I understand that this isn't necessarily a technical problem. It could be resolved by adding a little information about the circumstances under which 'Observe DOM changes' should be enabled by the user. But I don't know what the tag's original requirements were. The inconsistent behavior I desribed here seems undesirable to me though.
